File: src/main/java/com/fiseq1/simuladormetro/utils/Pseudoaleatorio.java
package com.fiseq1.simuladormetro.utils;

import com.fiseq1.simuladormetro.Controllers.SimuladorController;

/**
 * Clase generadora de números pseudoaleatorios basada en cálculos matemáticos,
 * manipulación de cadenas y generación congruencial lineal. Su rango de salida
 * se ajusta dinámicamente al valor de {@code PASAJEROS_RESTANTES}.
 * <p>
 * Esta clase forma parte del sistema de simulación de abordaje de pasajeros por estación
 * y garantiza que el valor generado nunca supere el límite restante de pasajeros.
 */
public class Pseudoaleatorio {

    private final double ALTO;
    private final double ANCHO;
    private final int valorAnterior;
    private String arrAscii = "";

    /**
     * Crea una nueva instancia del generador pseudoaleatorio.
     *
     * @param alto          Altura utilizada en los cálculos.
     * @param ancho         Ancho utilizado en los cálculos.
     * @param valorAnterior Influencia previa para generar la nueva semilla.
     */
    public Pseudoaleatorio(double alto, double ancho, int valorAnterior) {
        this.ALTO = alto;
        this.ANCHO = ancho;
        this.valorAnterior = valorAnterior;
    }

    /**
     * Ejecuta el proceso completo de generación de un número pseudoaleatorio.
     *
     * @return Valor entero generado, en el rango (1, PASAJEROS_RESTANTES).
     */
    public int generar() {
        return calcularPseudoaleatorio();
    }

    /**
     * Ejecuta internamente todo el algoritmo, desde la base matemática hasta la
     * aplicación del generador congruencial.
     *
     * @return Valor entero dentro del rango de pasajeros restantes.
     */
    private int calcularPseudoaleatorio() {
        int totalRestantes = SimuladorController.PASAJEROS_RESTANTES;

        if (totalRestantes <= 0) {
            return 1;
        }

        double PI = Math.PI;
        double base = ((ALTO * PI) + (ANCHO * PI)) * valorAnterior;
        String anchoAlto = String.valueOf(base);

        int[] asciiTemp = new int[anchoAlto.length() * 2];
        int count = 0;
        for (int i = 0; i < anchoAlto.length(); i++) {
            asciiTemp[count++] = (int) anchoAlto.charAt(i);
            asciiTemp[count++] = (int) ' ';
        }

        int tamArreglo = count / 2;
        int[] listaAscii = new int[tamArreglo];
        for (int i = 0; i < tamArreglo; i++) {
            String chunk = "" + (char) asciiTemp[i * 2] + (char) asciiTemp[i * 2 + 1];
            try {
                listaAscii[i] = Integer.parseInt(chunk.trim());
            } catch (NumberFormatException e) {
                listaAscii[i] = 32;
            }
        }

        int[] arreglo = burbujaCiega(listaAscii);

        String[] arregloStr = new String[arreglo.length];
        for (int i = 0; i < arreglo.length; i++) {
            arregloStr[i] = String.valueOf(arreglo[i]);
        }

        int semilla = generarSemilla(arregloStr);
        int[] lista = desordenControlado(semilla);
        return generadorCongruenciaLineal(semilla, lista);
    }

    /**
     * Intercambia pares adyacentes del arreglo una sola vez (ordenamiento ciego).
     *
     * @param arr Arreglo de enteros a modificar.
     * @return El mismo arreglo, pero con pares intercambiados una vez.
     */
    private int[] burbujaCiega(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            int temp = arr[i];
            arr[i] = arr[i + 1];
            arr[i + 1] = temp;
        }
        return arr;
    }

    /**
     * Genera una semilla a partir de un arreglo de cadenas numéricas,
     * sumando sus valores y agregando el valor anterior como influencia.
     *
     * @param arreglo Arreglo de cadenas que representan enteros.
     * @return Semilla entera resultante.
     */
    private int generarSemilla(String[] arreglo) {
        int semilla = 0;
        for (String s : arreglo) {
            try {
                semilla += Integer.parseInt(s);
            } catch (NumberFormatException e) {
                semilla += 32;
            }
        }
        semilla += valorAnterior;
        return semilla;
    }

    /**
     * Desordena una lista de enteros con base en una semilla,
     * simulando aleatoriedad determinista.
     *
     * @param semilla Valor que determina el patrón de mezcla.
     * @return Lista desordenada de enteros.
     */
    private int[] desordenControlado(int semilla) {
        int total = SimuladorController.PASAJEROS_RESTANTES;

        if (total <= 0) return new int[]{1};
        if (total == 1) return new int[]{0};

        int[] lista = new int[total];
        for (int i = 0; i < total; i++) {
            lista[i] = i;
        }

        for (int i = 0; i < total; i++) {
            int pos = (i * i + semilla) % total;
            if (pos == i || pos < 0 || pos >= total) continue;

            int temp = lista[i];
            lista[i] = lista[pos];
            lista[pos] = temp;
        }

        return lista;
    }

    /**
     * Aplica un generador congruencial lineal a partir de una semilla y una lista de enteros.
     * Ajusta los coeficientes dinámicamente según el tamaño actual de la lista.
     *
     * @param semilla Valor base para el generador.
     * @param lista   Lista de enteros con influencia en los coeficientes.
     * @return Número pseudoaleatorio resultante dentro del rango permitido.
     */
    private int generadorCongruenciaLineal(int semilla, int[] lista) {
        int total = SimuladorController.PASAJEROS_RESTANTES;
        if (total <= 0 || lista.length == 0) return 1;

        int indexA = Math.min(total / 4, lista.length - 1);
        int indexC = Math.min(total / 6, lista.length - 1);
        int indexX = Math.min(total / 3, lista.length - 1);

        int a = 92 + lista[indexA];
        int c = 164 + lista[indexC];
        int x = semilla * lista[indexX];

        x = (a * x + c) % (total + 1);
        x = (a * x + c) % (total + 1);

        if (x == 0) x = 1;
        return x;
    }

    /**
     * Devuelve la cadena ASCII generada internamente (si se usara).
     *
     * @return Cadena con valores ASCII concatenados.
     */
    public String getArrAscii() {
        return arrAscii;
    }

    /**
     * Establece manualmente la cadena ASCII interna.
     *
     * @param arrAscii Nueva cadena a asignar.
     */
    public void setArrAscii(String arrAscii) {
        this.arrAscii = arrAscii;
    }
}

File: src/main/java/com/fiseq1/simuladormetro/utils/Pseudoaleatorio.kt
package com.fiseq1.simuladormetro.utils

import com.fiseq1.simuladormetro.Controllers.SimuladorController

/**
 * Clase generadora de números pseudoaleatorios basada en transformaciones matemáticas
 * con influencia del valor anterior generado y dimensiones proporcionadas.
 *
 * El número generado está limitado a un rango de 1 a [TOTAL_PASAJEROS] (40 por defecto).
 * Combina manipulación de cadenas ASCII, ordenamientos simples, una semilla influenciada por el
 * valor anterior y un generador congruencial lineal.
 *
 * @param alto Altura usada como parte del cálculo base para la generación.
 * @param ancho Ancho usado como parte del cálculo base para la generación.
 * @param valorAnterior Valor numérico usado como influencia en la semilla. Por defecto es 0.
 */
class Pseudoaleatorio(var alto: Double, var ancho: Double, var valorAnterior: Int = 0) {

    /** Valor constante de altura para uso interno. */
    val ALTO = alto

    /** Valor constante de ancho para uso interno. */
    val ANCHO = ancho

    /** Valor de PI utilizado en cálculos base. */
    val PI = Math.PI

    /** Número máximo de pasajeros (límite superior del número aleatorio). */
    val TOTAL_PASAJEROS = SimuladorController.TOTAL_PASAJEROS

    /** Cadena intermedia con valores ASCII (actualmente sin uso externo). */
    var arrAscii: String = ""

    /** Resultado generado internamente. */
    private var resultado: Int = 0

    /**
     * Ejecuta todo el proceso de generación del número pseudoaleatorio:
     * - Calcula una base usando PI, alto, ancho y el valor anterior.
     * - Convierte la base en una cadena de caracteres ASCII.
     * - Aplica ordenamiento básico (burbuja ciega).
     * - Genera una semilla.
     * - Desordena una lista de forma controlada.
     * - Aplica un generador congruencial lineal.
     *
     * @return Número pseudoaleatorio resultante en el rango (1, TOTAL_PASAJEROS).
     */
    private fun calcularPseudoaleatorio(): Int {
        val base = ((ALTO * PI) + (ANCHO * PI)) * valorAnterior
        val anchoAlto = "$base"
        println("Base modificada con valorAnterior ($valorAnterior): $anchoAlto")

        val cadenaAscii = StringBuilder()
        for (char in anchoAlto) {
            cadenaAscii.append(char.code).append(' '.code)
        }

        val cadenaAsciiStr = cadenaAscii.toString()
        val listaAscii = cadenaAsciiStr.chunked(2).map { it.toInt() }.toMutableList()

        val arreglo = burbujaCiega(listaAscii)
        val arregloStr = arreglo.map { it.toString() }.toTypedArray()
        val semilla = generarSemilla(arregloStr)
        val lista = desordenControlado(semilla)
        resultado = generadorCongruenciaLineal(semilla, lista)
        return resultado
    }

    /**
     * Aplica un ordenamiento mínimo (burbuja ciega) intercambiando pares adyacentes una sola vez.
     *
     * @param arr Lista mutable de enteros a reordenar.
     * @return Lista modificada tras la operación.
     */
    fun burbujaCiega(arr: MutableList<Int>): MutableList<Int> {
        for (i in 0 until arr.size - 1) {
            val temp = arr[i]
            arr[i] = arr[i + 1]
            arr[i + 1] = temp
        }
        return arr
    }

    /**
     * Genera una semilla entera sumando los valores ASCII de los elementos del arreglo
     * y añadiendo el [valorAnterior] como parte del cálculo.
     *
     * @param arreglo Arreglo de cadenas numéricas derivadas de caracteres ASCII.
     * @return Valor entero que servirá como semilla para el algoritmo.
     */
    fun generarSemilla(arreglo: Array<String>): Int {
        var semilla = 0
        for (car in arreglo) {
            val codigo = car.toIntOrNull() ?: 32
            semilla += codigo
        }
        semilla += valorAnterior
        return semilla
    }

    /**
     * Reorganiza una lista de enteros de forma controlada con base en la semilla,
     * simulando una mezcla pseudoaleatoria pero determinista.
     *
     * @param semilla Valor entero que influye en el patrón de desordenamiento.
     * @return Lista de enteros reorganizada.
     */
    fun desordenControlado(semilla: Int): IntArray {
        val lista = IntArray(TOTAL_PASAJEROS) { it + 1 }
        for (i in 0 until TOTAL_PASAJEROS) {
            lista[i] = i
        }
        for (i in 0 until TOTAL_PASAJEROS) {
            var pos = (i * i + semilla) % TOTAL_PASAJEROS
            if (pos == 0) pos = 1
            val temp = lista[i]
            lista[i] = lista[pos]
            lista[pos] = temp
        }
        return lista
    }

    /**
     * Generador congruencial lineal que usa la semilla y una lista
     * para obtener un número dentro del rango permitido.
     *
     * @param semilla Valor inicial del generador.
     * @param lista Lista de enteros usada para calcular coeficientes dinámicos.
     * @return Número pseudoaleatorio en el rango (1, TOTAL_PASAJEROS).
     */
    fun generadorCongruenciaLineal(semilla: Int, lista: IntArray): Int {
        val a = 92 + lista[TOTAL_PASAJEROS / 4]
        val c = 164 + lista[TOTAL_PASAJEROS / 6]
        var x = semilla * lista[TOTAL_PASAJEROS / 3]
        x = (a * x + c) % (TOTAL_PASAJEROS + 1)
        x = (a * x + c) % (TOTAL_PASAJEROS + 1)
        if (x == 0) x = 1
        return x
    }

    /**
     * Método público que permite generar el número pseudoaleatorio.
     *
     * @return Valor generado tras aplicar todos los algoritmos internos.
     */
    fun generar(): Int {
        return calcularPseudoaleatorio()
    }
}
